### ISSUE TITLE
Target tokens for loss computation

### ISSUE THREAD

**Issue by yukara-ikemiya** [valeoai/Maskgit-pytorch#3]
Hi, I'd like to qustion about a loss computation part.
This repository (and the original repository?) compute cross-emtropy loss with entire groud-truth tokens.
This implies that the model learns to predict 'known (unmasked)' tokens as well, which is relatively easy to estimate.
As a result, the training may exhibit a strong bias towards the known tokens.
https://github.com/valeoai/Maskgit-pytorch/blob/d2ba643ccd915d81895089e87d287445a16416a1/Trainer/vit.py#L191
Intuitively thinking, in this case, a model firstly ignore the loss of 'masked' tokens and the loss of known tokens would be drastically descreased at the beginning of training. 

I think there is another option of masking the known position in the target tokens, which results in forcing a model to predict only the unknown (masked) tokens (as same as the approach taken in the following repository).
https://github.com/lucidrains/muse-maskgit-pytorch/blob/main/muse_maskgit_pytorch/muse_maskgit_pytorch.py#L680

I'd like to know if you have any insights on the following two points regarding this.

1. Which approach tends to yield better learning outcomes: masking the target or not masking the target?
2. Could you share the loss curve through epochs during training so that we can confirm if our training is going well?

Thank you for considering my request!
Best regards,
Yukara

**Comment by llvictorll** [comment]
Hi, 

Yes, there is a difference between my repo and the one from lucidrains. I do not know if the official repository use the same strategy since they do not release the training script. 

1. I think that keeping the prediction on the 'known (unmasked)' token will help to quickly see if the network is learning or not, for debugging purposes. I do not think that the prediction will be biased towards the 'known (unmasked)' tokens because the masking ratio is high (i.e. arccos scheduler) and the loss for 'known (unmasked)' tokens will be very low very quickly.

2. Here is the learning curve from the shared model. It shows some picking in the gradient curve (epoch 8, 33, and 85) that might be eliminated by more drastic gradient_norm_clipping.  

![learning_curve](https://github.com/valeoai/Maskgit-pytorch/assets/32369153/f03c8720-505b-4743-b953-e7857688ef90)

**Comment by yukara-ikemiya** [comment]
Thank you for the respose. The loss curve plot is very helpful for me.
